pkg/store/storepb/matchers_cache.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storepb
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const CachedMatcherTTL = 5 * time.Minute
+
+type NewItemFunc func(matcher LabelMatcher) (*labels.Matcher, error)
+
+type itemExpiration struct {
+	mu        sync.Mutex
+	expiresAt time.Time
+}
+
+func (e *itemExpiration) setExpiration(t time.Time) {
+	e.mu.Lock()
+	e.expiresAt = t
+	e.mu.Unlock()
+}
+
+func newExpiration(expiresAt time.Time) *itemExpiration {
+	return &itemExpiration{
+		expiresAt: expiresAt,
+	}
+}
+
+type MatchersCache struct {
+	reg *prometheus.Registry
+	now func() time.Time
+
+	mu       sync.RWMutex
+	ttl      time.Duration
+	itemTTLs map[LabelMatcher]*itemExpiration
+	cache    map[LabelMatcher]*labels.Matcher
```

Comment:
How could we ensure that this doesn't grow endlessly? For example, with this, it is quite easy to do a DoS attack by just sending a bunch of random queries with different matchers.

Reply:
That's a good point. I also noticed that memory went up during my experiment, so it seems like we traded CPU for memory. 

I am not sure if this is a good long-term optimization so I will close the PR for now.

Reply:
Can we have a size based cache instead?